backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java

```diff
@@ -580,4 +613,46 @@ private SyncToParentStatus deriveSyncToParentStatus(
       }
     }
   }
+
+  private static class InferUpstreamForLocalBranchAux extends BaseAux {
+
+    InferUpstreamForLocalBranchAux(IGitCoreRepository gitCoreRepository) throws GitCoreException {
+      super(gitCoreRepository);
+    }
+
+    Option<String> inferUpstreamForLocalBranch(
```

Comment:
This is the actual new piece of code.

Reply:
It's in fact very similar to fork point algorithm: inferred upstream of `X` is (roughly) just one of the branches on whose reflog the fork point for `X` has been found